src/zafClient.re

```diff
@@ -0,0 +1,174 @@
+type product =
+  | Support
+  | Chat;
+
+type location =
+  | Background
+  | Modal
+  | NavBar
+  | NewTicketSidebar
+  | OrganizationSidebar
+  | TicketEditor
+  | TicketSidebar
+  | TopBar
+  | UserSidebar;
+
+type subdomain =
+  | Subdomain(string);
+
+type account = {subdomain: option(subdomain)};
+
+type context = {
+  instanceGuid: Utils.guid,
+  product,
+  account: option(account),
+  location,
+  ticketId: option(int)
+};
+
+type requestOptions = {
+  url: Utils.url,
+  secure: bool,
+  _type: Utils.requestMethod
+};
+
+type hashbackStatus = {
+  hasOauthToken: bool,
+  subdomain: option(subdomain)
+};
+
+module Decode = {
+  let product = productString =>
+    switch productString {
+    | "support" => Support
+    | "chat" => Chat
+    | e => raise(Failure("Unknown product: " ++ e))
+    };
+  let location = locationString =>
+    switch locationString {
+    | "background" => Background
+    | "modal" => Modal
+    | "nav_bar" => NavBar
+    | "new_ticket_sidebar" => NewTicketSidebar
+    | "organization_sidebar" => OrganizationSidebar
+    | "ticket_editor" => TicketEditor
+    | "ticket_sidebar" => TicketSidebar
+    | "top_bar" => TopBar
+    | "user_sidebar" => UserSidebar
+    | e => raise(Failure("Unknown product: " ++ e))
+    };
+  let subdomain = maybeSubdomainString =>
+    switch maybeSubdomainString {
+    | Some(subdomainString) => Some(Subdomain(subdomainString))
+    | None => None
+    };
+  let account = json =>
+    Json.Decode.{
+      subdomain: json |> optional(field("subdomain", string)) |> subdomain
+    };
+  let context = json =>
+    Json.Decode.{
+      instanceGuid: json |> field("instanceGuid", string) |> Utils.createGuid,
+      product: json |> field("product", string) |> product,
+      account: json |> optional(field("account", account)),
+      location: json |> field("location", string) |> location,
+      ticketId: json |> optional(field("ticketId", int))
+    };
+  let hashbackStatus = json =>
+    Json.Decode.{
+      hasOauthToken: json |> field("hasOauthToken", bool),
+      subdomain: json |> optional(field("subdomain", string)) |> subdomain
+    };
+};
+
+type zafClient;
+
+[@bs.scope "ZAFClient"] [@bs.val] external init : unit => zafClient = "";
+
+[@bs.send]
+external requestContext : zafClient => Js.Promise.t(Js.Json.t) = "context";
+
+[@bs.send]
+external request :
+  (
+    zafClient,
+    {
+      .
+      "url": string,
+      "secure": Js.boolean,
+      "_type": string
+    }
+  ) =>
+  Js.Promise.t(Js.Json.t) =
+  "";
+
+[@bs.send]
+external on_app_registered :
+  (zafClient, [@bs.as "app.registered"] _, unit => unit) => unit =
+  "on";
```

Comment:
can you break this down for me?

What I get from it so far is "on_app_registered is a function that accepts a zafClient, a function that goes from unit => unit, and returns a unit. Also it has hard coded 'app.registered' as the first argument to zafClient.on". Is that all right? so it compiles to

```javascript
const on_app_registred = (x, y) => x.on('app.registered', y);
```

but my real question is why the `_` after `[@bs.as "app.registered"]`

Reply:
I lifted it from https://bucklescript.github.io/docs/en/function.html#fixed-arguments

I don't actually know what the `_` does :)